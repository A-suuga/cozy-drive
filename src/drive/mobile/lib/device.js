// platform
export const isCordova = () => window.cordova !== undefined
export const getPlatformId = () =>
  isCordova() ? window.cordova.platformId : undefined
const isPlatform = platform => getPlatformId() === platform
export const isIos = () => isPlatform('ios')
export const isAndroid = () => isPlatform('android')

// device
const hasCordovaDeviceNamePlugin = () =>
  isCordova() && window.cordova.plugins.deviceName !== undefined
export const getDeviceName = () =>
  hasCordovaDeviceNamePlugin()
    ? window.cordova.plugins.deviceName.name
    : undefined
