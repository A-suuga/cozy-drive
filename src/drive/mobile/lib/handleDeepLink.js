const PROTOCOL = 'cozydrive://'
const RX = new RegExp('^' + PROTOCOL)

export default history => url => {
  const stripped = url.replace(RX, '')
  history.push(stripped)
}
