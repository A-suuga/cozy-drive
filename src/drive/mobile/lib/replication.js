/* global cozy */
import { clientRevokedMsg } from './cozy-helper'

export const startReplication = (firstReplication, firstReplicationFinished, refreshFolder, revokeClient) => {
  if (firstReplication) {
    startRepeatedReplication(refreshFolder, revokeClient)
  } else {
    startFirstReplication(firstReplicationFinished, refreshFolder, revokeClient)
  }
}

const startRepeatedReplication = (refreshFolder, revokeClient) => {
  const options = {
    onError: onError(revokeClient),
    onComplete: result => {
      if (result.docs_written !== 0) {
        refreshFolder()
      }
    }
  }
  cozy.client.offline.startRepeatedReplication('io.cozy.files', 15, options)
}

const startFirstReplication = (firstReplicationFinished, refreshFolder, revokeClient) => {
  const options = {
    onError: onError(revokeClient),
    onComplete: () => {
      firstReplicationFinished()
      startRepeatedReplication(refreshFolder, revokeClient)
    }
  }
  cozy.client.offline.replicateFromCozy('io.cozy.files', options).then(() => {
    console.log('End of Replication')
  })
}

export const onError = (revokeClient) => (err) => {
  if (err.message === clientRevokedMsg || err.error === 'code=400, message=Invalid JWT token') {
    console.warn('Your device is no more connected to your server')
    revokeClient()
  } else if (err.message === 'ETIMEDOUT') {
    console.log('timeout')
  } else {
    console.warn(err)
  }
}
