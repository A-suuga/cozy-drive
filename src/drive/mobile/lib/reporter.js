/* global __SENTRY_TOKEN__, __DEVELOPMENT__ */
import Raven from 'raven-js'

let isEnabled = false

const getAnalyticsUrl = () => {
  if (typeof __SENTRY_TOKEN__ === 'undefined') {
    return ''
  }

  const PROD_ANALYTICS_URL = `https://${__SENTRY_TOKEN__}@sentry.cozycloud.cc/6`
  const DEV_ANALYTICS_URL = `https://${__SENTRY_TOKEN__}@sentry.cozycloud.cc/2`
  return __DEVELOPMENT__ ? DEV_ANALYTICS_URL : PROD_ANALYTICS_URL
}

export const ANALYTICS_URL = getAnalyticsUrl()

export const getReporterConfiguration = () => ({
  shouldSendCallback: () => isEnabled,
  environment: __DEVELOPMENT__ ? 'development' : 'production'
})

export const configureReporter = enable => {
  isEnabled = enable
  Raven.config(ANALYTICS_URL, getReporterConfiguration()).install()
}

export const logException = err => {
  return new Promise(resolve => {
    Raven.captureException(err)
    console.warn('Raven is recording exception')
    console.error(err)
    resolve()
  })
}

const logMessage = (message, serverUrl, level = 'info', force) => {
  return new Promise(resolve => {
    if (force) {
      configureReporter(true)
    }
    Raven.setUserContext = {
      url: serverUrl
    }
    Raven.captureMessage(`[${serverUrl}] ${message}`, {
      level
    })
    if (force) {
      configureReporter(isEnabled)
    }
    resolve()
  })
}

export const logInfo = (message, serverUrl, force = false) =>
  logMessage(message, serverUrl, 'info', force)
