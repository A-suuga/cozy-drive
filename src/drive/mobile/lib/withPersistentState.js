import localforage from 'localforage'

// The withPersistentState HOC automatically saves the wrappped components state and restores it when the component is mounted again.
// Since the saved state will not be available immediately when the wrapped component is mounted, a `componentStateRestored` hook will be called once the state is restored
export const withPersistentState = (WrappedComponent, persistenceKey) => {
  class WithPersistentState extends WrappedComponent {
    async componentWillMount() {
      const persistedState = await localforage.getItem(persistenceKey)
      this.setState(
        prevState => persistedState || prevState,
        () => {
          if (super.componentStateRestored) super.componentStateRestored()
        }
      )
    }

    componentWillUpdate(nextProps, nextState) {
      if (super.componentWillUpdate)
        super.componentWillUpdate(nextProps, nextState)
      localforage.setItem(persistenceKey, nextState)
    }
  }
  WithPersistentState.displayName = `WithPersistentState(${WrappedComponent.displayName ||
    WrappedComponent.name ||
    'Component'})`
  return WithPersistentState
}
export default withPersistentState
