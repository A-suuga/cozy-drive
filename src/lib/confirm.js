import React from 'react'
import ReactDOM from 'react-dom'

const confirm = component => new Promise((resolve, reject) => {
  const wrapper = document.body.appendChild(document.createElement('div'))

  const abort = () => {
    ReactDOM.unmountComponentAtNode(wrapper)
    reject()
  }

  const confirm = () => {
    ReactDOM.unmountComponentAtNode(wrapper)
    resolve()
  }

  ReactDOM.render(React.cloneElement(component, { confirm, abort }), wrapper)
})

export default confirm
