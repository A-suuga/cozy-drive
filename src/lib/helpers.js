export const getPhotosByMonth = photos => {
  let sections = {}
  photos.forEach(p => {
    // here we want to get an object whose keys are months in a l10able format
    // so we only keep the year and month part of the date
    const month = p.metadata.datetime.slice(0, 7) + '-01T00:00'
    /* istanbul ignore else */
    if (!sections.hasOwnProperty(month)) {
      sections[month] = []
    }
    sections[month].push(p)
  })
  // we need to sort the months here because when new photos are uploaded, they
  // are inserted on top of the list, and months can become unordered
  const sortedMonths = Object.keys(sections).sort(
    (a, b) => new Date(a).getTime() - new Date(b).getTime()
  ).reverse()
  
  return sortedMonths.map(month => {
    return {
      title: month,
      photos: sections[month]
    }
  })
}
