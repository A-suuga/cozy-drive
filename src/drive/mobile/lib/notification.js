/* global cordova */
import { isCordova } from './device'
const hasPlugin = () =>
  isCordova() && cordova.plugins.notification.local !== undefined

export const scheduleNotification = options => {
  if (hasPlugin()) cordova.plugins.notification.local.schedule(options)
}
