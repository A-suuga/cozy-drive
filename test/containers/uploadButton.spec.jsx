'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { mockT } from '../lib/I18n'
import Connected, { UploadButton, mapDispatchToProps } from '../../src/containers/UploadButton'

import thunk from 'redux-thunk'
import configureMockStore from 'redux-mock-store'
const middlewares = [ thunk ]
const mockStore = configureMockStore(middlewares)

describe('UploadButton component', () => {
  it('should be rendered correctly without store', () => {
    // render
    const component = shallow(
      <UploadButton
        t={mockT}
        uploadPhotos={(photosArray) => photosArray}
      />
    ).node
    // test componentWillReceiveProps
    expect(component).toMatchSnapshot()
  })

  it('should be handled correctly the input onChange for uploadPhotos', () => {
    // set
    const uploadPhotosMock = jest.fn()
    // render
    const component = shallow(
      <UploadButton
        t={mockT}
        uploadPhotos={uploadPhotosMock}
      />
    )
    // simulate an onChange event
    const e = {
      target: {
        files: [
          {
            name: 'MyImage.jpg',
            size: 128392
          }
        ]
      }
    }
    component.node.props.children[1].props.onChange(e)

    expect(uploadPhotosMock.mock.calls.length).toBe[1]
    expect(component).toMatchSnapshot()
  })

  it('should be rendered correctly with store', () => {
    const store = mockStore({})
    const component = shallow(
      <Connected
        t={mockT}
        store={store}
        uploadPhotos={(photosArray) => photosArray}
      />
    )
    expect(component.shallow().node).toMatchSnapshot()
  })

  it('should use a correct a mapDispatchToProps with correct dispatch calls', () => {
    // set
    const dispatchMock = jest.fn()
    const photo = {
      name: 'monImage.jpg',
      size: 178298
    }
    const nextProps = mapDispatchToProps(dispatchMock, null)

    // act
    nextProps.uploadPhotos(photo)

    // assert
    // onFirstFetch call an action using dispatch
    expect(dispatchMock.mock.calls.length).toBe(1)
  })
})
