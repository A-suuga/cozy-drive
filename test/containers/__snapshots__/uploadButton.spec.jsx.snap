exports[`UploadButton component should be handled correctly the input onChange for uploadPhotos 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <label
    className="coz-btn coz-btn--regular coz-btn--upload"
    role="button"
    style={
        Object {
            "overflow": "hidden",
            "position": "relative",
          }
    }>
    Upload
    <input
        accept="image/*"
        multiple={true}
        onChange={[Function]}
        style={
            Object {
                "height": "100%",
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": "100%",
                "zIndex": 1,
              }
        }
        type="file" />
</label>,
  "nodes": Array [
    <label
      className="coz-btn coz-btn--regular coz-btn--upload"
      role="button"
      style={
            Object {
                  "overflow": "hidden",
                  "position": "relative",
                }
      }>
      Upload
      <input
            accept="image/*"
            multiple={true}
            onChange={[Function]}
            style={
                  Object {
                        "height": "100%",
                        "left": 0,
                        "opacity": 0,
                        "position": "absolute",
                        "top": 0,
                        "width": "100%",
                        "zIndex": 1,
                      }
            }
            type="file" />
</label>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Unknown
        t={[Function]}
        uploadPhotos={[Function]} />,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "t": [Function],
          "uploadPhotos": [Function],
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <label
          className="coz-btn coz-btn--regular coz-btn--upload"
          role="button"
          style={
                    Object {
                              "overflow": "hidden",
                              "position": "relative",
                            }
          }>
          Upload
          <input
                    accept="image/*"
                    multiple={true}
                    onChange={[Function]}
                    style={
                              Object {
                                        "height": "100%",
                                        "left": 0,
                                        "opacity": 0,
                                        "position": "absolute",
                                        "top": 0,
                                        "width": "100%",
                                        "zIndex": 1,
                                      }
                    }
                    type="file" />
</label>,
        "_debugID": 4,
        "_renderedOutput": <label
          className="coz-btn coz-btn--regular coz-btn--upload"
          role="button"
          style={
                    Object {
                              "overflow": "hidden",
                              "position": "relative",
                            }
          }>
          Upload
          <input
                    accept="image/*"
                    multiple={true}
                    onChange={[Function]}
                    style={
                              Object {
                                        "height": "100%",
                                        "left": 0,
                                        "opacity": 0,
                                        "position": "absolute",
                                        "top": 0,
                                        "width": "100%",
                                        "zIndex": 1,
                                      }
                    }
                    type="file" />
</label>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Unknown
    t={[Function]}
    uploadPhotos={[Function]} />,
}
`;

exports[`UploadButton component should be rendered correctly with store 1`] = `
<label
  className="coz-btn coz-btn--regular coz-btn--upload"
  role="button"
  style={
    Object {
      "overflow": "hidden",
      "position": "relative",
    }
  }>
  Upload
  <input
    accept="image/*"
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "height": "100%",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 1,
      }
    }
    type="file" />
</label>
`;

exports[`UploadButton component should be rendered correctly without store 1`] = `
<label
  className="coz-btn coz-btn--regular coz-btn--upload"
  role="button"
  style={
    Object {
      "overflow": "hidden",
      "position": "relative",
    }
  }>
  Upload
  <input
    accept="image/*"
    multiple={true}
    onChange={[Function]}
    style={
      Object {
        "height": "100%",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "width": "100%",
        "zIndex": 1,
      }
    }
    type="file" />
</label>
`;
